### ISSUE TITLE
Valgrind Reports leak for any loop which had an ev_signal_init run against it.

### ISSUE THREAD

**Issue by coreyfarrell** [libuv/libuv#298]
I test every application I write with valgrind, while testing libuv with some basic daemon code I found a leak.  Sample code to produce this leak is at https://gist.github.com/coreyfarrell/bc8d6f9e4cb6983e2468.  I attempted to find a fix but was unsuccessful.

**Comment by saghul** [comment]
There is global pipe which is used to dispatch signals and it's currently
never closed.

Now, it's not really a big deal since it will be closed then the
application ends, but if you can come up with a fix, it would be welcome.

Thanks!

**Comment by coreyfarrell** [comment]
I'm not sure I understand, you say it's global but it's actually per loop.  If I start and stop 50 loops, I get 50 leaks.

I checked into this with GDB, in every case uv_loop_close return's EV_BUSY due to &sig still being listed in loop->handle_queue with sig.flags == UV__HANDLE_REF.  I also tested further by adding a timer that fired every 100ms, the first time it fired would run uv_signal_stop(&sig) and 10th time it fired would stop the timer.  This had no effect on the non-completion of uv_loop_close or the leak.  Since 1 second is an eternity in computing time I am fairly confident to say uv_loop_close can never complete once a signal is initialized against the loop.

So it seems to me that once a signal handler is added to a loop it can never be removed, and prevents the loop cleanup from completing.  As you say not a big deal in a situation where a single loop is being used, but this would be a big deal if multiple/temporary loops are being used.

This also poses problems if 'uv_signal_t sig;' is malloc'ed (for example as part of another structure).  At what point is it safe to free the location where sig is stored?  I did an additional test:
```
int i2 = 0;
while (++i2 < 100) {
	uv_signal_t *s2;
	s2 = malloc(sizeof(*s2));
	uv_signal_init(&l, s2);
	uv_signal_start(s2, signal_cb, SIGINT);
	uv_signal_stop(s2);
	free(s2);
}
```

As expected this produced a number of valgrind memory access errors.  uv_signal_init was performing writes to memory freed in previous iterations of this loop.  It also caused memory access after free during ev_loop_close.

**Comment by bnoordhuis** [comment]
> At what point is it safe to free the location where sig is stored?

Inside the close callback that you pass as an argument to uv_close().

**Comment by saghul** [comment]
There is a global pipe, but of course, of you don't close all handles fds
will remain open. You can use uv_walk to iterate all handles, close them as
Ben suggested and then call uv_loop_close. If it returns success all loop
resources were closed. (Except that global pipe, that is)

**Comment by coreyfarrell** [comment]
I've found the problem.  If you run uv_signal_stop after the loop has completed, uv__run_closing_handles is never run again.  Adding an additional call to uv_run after stopping the signal watcher worked around the issue (valgrind reports 0 bytes leaked memory).  I believe the correct fix is to call uv__run_closing_handles(loop) at the start of uv_loop_close.  I've posted pull request #299 with this fix.

Since Github doesn't allow conversion from issue to pull request I am closing this issue.
